server/src/main/java/org/opensearch/cluster/metadata/Metadata.java

```diff
@@ -1139,8 +1157,17 @@ public Builder(Metadata metadata) {
             this.hashesOfConsistentSettings = metadata.hashesOfConsistentSettings;
             this.version = metadata.version;
             this.indices = new HashMap<>(metadata.indices);
+            this.indicesPreviousState = new HashMap<>(metadata.indices); // required for comparing with updated indices
             this.templates = new HashMap<>(metadata.templates);
             this.customs = new HashMap<>(metadata.customs);
+            this.dataStreamPreviousMetadata = (DataStreamMetadata) metadata.customs.get(DataStreamMetadata.TYPE);
+            this.indicesLookup = new TreeMap<>(metadata.indicesLookup);
+            this.allIndices = Arrays.copyOf(metadata.allIndices, metadata.allIndices.length);
+            this.visibleIndices = Arrays.copyOf(metadata.visibleIndices, metadata.visibleIndices.length);
+            this.allOpenIndices = Arrays.copyOf(metadata.allOpenIndices, metadata.allOpenIndices.length);
+            this.visibleOpenIndices = Arrays.copyOf(metadata.visibleOpenIndices, metadata.visibleOpenIndices.length);
+            this.allClosedIndices = Arrays.copyOf(metadata.allClosedIndices, metadata.allClosedIndices.length);
+            this.visibleClosedIndices = Arrays.copyOf(metadata.visibleClosedIndices, metadata.visibleClosedIndices.length);
```

Comment:
There is an empty Builder constructor as well. You would need null checks when accessing this previous metadata object. Yes, there is definitely overhead in creating these array ahead of time if metadata actually changes and they have to be recomputed. The downsides of keeping the previousMetadata object in private variable : 
1. it will not be garbage collected immediately, that should be ok.
2. On the immutability ```Metadata``` is immutable but internally all the fields it stores like indices are mutable and if they are updated by mistake, it can corrupt the current cluster state.
Generally, it is better for the copyConstructor to read the fields it cares, initialize the new object and release the input object.

Reply:
I cross checked ```indices```, ```custom```, ```templates``` are ```Collections.unmodifiableMap```

Reply:
Yes, we will/should not expose previousMetadata via any method. Instead of storing all these arrays, it will be a cleaner code design to have the previous Metadata just being referenced.

Also, for these array objects: visibleIndices, allOpenIndices....., I see their usage is [mostly](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/cluster/metadata/IndexNameExpressionResolver.java#L1175) as lists.

It might be better to store these objects as `Collections.unmodifiableList` itself as storing these objects as arrays possess mutability issues. Although, we may have to scope this change outside this PR, as their relevant XContent parser methods/tests might also required changes + refactoring at other places.

Reply:
@shwetathareja I have made changes as I was suggesting to store reference of previous Metadata. The code gets a lot cleaner without heavy lifting the builder method also.